design/multi-checkout.md

```diff
@@ -0,0 +1,181 @@
+- Alter our existing behaviors when there are 0 or 1 `checkout` steps
+- Design for multiple repository triggers (look for a sibling document shortly)
+- Change our security posture for additional Azure Repos (other features may address this in the future)
+
+## Customer scenarios
+
+Customers often have source, tools, scripts, or other collateral stored in a secondary repo.
+Even a big mono-repo like the Azure DevOps codebase has adjunct repositories that could be used in a CI/CD pipeline.
+Especially for teams working with microservices, it's common to require additional sidecar repos containing shared infrastructure or SDKs.
+
+On the other hand, we have lots of customers using just a single repository in their pipeline, and they plan to stay that way.
+Any features we build must respect the investment those customers have made in their pipelines.
+Therefore we have this stake in the ground:
+
+> We are in "multi-checkout" when the customer has explicitly asked for multiple repositories to be checked out.
+> Zero checkout steps yields the same behavior as today (checkout `self` in a normal job, no checkout in a `deployment` job).
+> One checkout (whether it's `self` or not) preserves existing single-repo behavior.
+
+## Multi-checkout mode
+
+_(Note: I'm just calling it a "mode" for purposes of discussion here. We won't document it as a "mode".)_
+
+The agent will look at the job message and discover all `checkout` steps.
+
+### Zero or one checkout step
+0. If there are none at all, we'll preserve the existing default behavior, acting as if a `checkout: self` were the first step.
+0. If there is exactly one `checkout: none`, then we'll preserve the existing behavior: no repositories are synced or checked out.
+0. If there is exactly one `checkout` step that isn't `none`, then we'll use existing single-checkout behavior but check out that repository rather than `self`.
+
+### More than one checkout step
+0. If there are multiple checkout steps and one or more are `checkout: none`, this is an error.
+It's unclear what the user meant for us to do, so we fail with a clear message.
+0. Otherwise, we are in multi-checkout mode.
+
+### Behavior changes in multi-checkout mode
+
+#### Checkout directory
+By default, all repositories will be checked out as if they've been `git clone`d into the `Pipeline.Workspace` directory.
+Git uses the last part of the path, minus any trailing slashes, spaces, and the string `.git`.
+(We have YAML syntax for selecting a different path, and if the user specifies that, we'll use their name.)
+This changes the default directory for the `self` repo: it would be `$(Pipeline.Workspace)/s` if it were the only `checkout` step.
+
+#### System variables
+`Build.SourcesDirectory` points to the directory where the `self` repo was checked out whether in single- or multi-checkout mode.
```

Comment:
What will Build.SourcesDirectory be set to if self isn't one of the listed checkout values?

Reply:
Good question. Today if you `checkout: none`, we still create an `s/` directory and make that the `Build.SourcesDirectory`. I think the thing that breaks least is to do the same here. Thoughts @scottboehmer and @jpricket ?

(Someone offline suggested we always make it the directory that contains the YAML pipeline. But that won't work once we have repository triggers, since the repo with the YAML might not be checked out or even available.)

Reply:
Same behavior as the `checkout: none` case seems reasonable to me. Is there anything to prevent a user's yaml from claiming the `s/` path for a different repo or are we okay if they do that and end up with `Build.SourcesDirectory` pointing to whichever repo they put at that path?

Reply:
No, and there's also nothing really stopping a user from pointing to `a/`, `b/`, or `testResults/` either. I'm OK with the risk (and of course we'll doc). I should mention all that in the spec.

Reply:
Seems reasonable to me.